src/starkware_utils/commitment_tree/binary_fact_tree_node.rs

```diff
@@ -0,0 +1,322 @@
+            _storage: Default::default(),
+            _hash_function_type: Default::default(),
+            _leaf_fact: Default::default(),
+        }
+    }
+}
+
+impl<'test, S, H, LF, N> TreeTraverser<'test, S, H, LF> for BinaryFactTreeTraverser<'test, S, H, LF, N>
+where
+    S: Storage + Sync + Send,
+    H: HashFunctionType + Sync + Send,
+    LF: LeafFact<S, H> + Sync + Send,
+    N: BinaryFactTreeNode<S, H, LF> + Sync + 'test,
+{
+    type NodeType = BinaryFactTreeDiff<S, H, LF, N>;
+
+    async fn get_children(&mut self, node: &Self::NodeType) -> Result<Vec<Self::NodeType>, TreeError> {
+        if node.previous.is_leaf() {
+            // unwrap() on leaf_hash is guaranteed to be safe because of the check above
+            let previous = LF::get_or_fail(&self.ffc.storage, &node.previous.leaf_hash().unwrap()).await?;
+            let current = LF::get_or_fail(&self.ffc.storage, &node.current.leaf_hash().unwrap()).await?;
+            self.result.push(BinaryFactTreeNodeDiff::new(node.path, previous, current));
+            return Ok(vec![]);
+        }
+
+        let (previous_left, previous_right) = node.previous.get_children(self.ffc, self.facts).await?;
+        let (current_left, current_right) = node.current.get_children(self.ffc, self.facts).await?;
+
+        let mut children = vec![];
+        if previous_left != current_left {
+            // Shift left for the left child
+            let path = node.path << 1;
+            children.push(BinaryFactTreeDiff::new(path, previous_left, current_left));
+        }
+        if previous_right != current_right {
+            // Shift left and turn on the LSB bit for the right child
+            let path = (node.path << 1) + 1;
+            children.push(BinaryFactTreeDiff::new(path, previous_right, current_right));
+        }
+
+        Ok(children)
+    }
+}
+
+pub trait BinaryFactTreeNode<S, H, LF>: Sized + PartialEq + Sync + Send
+where
+    S: Storage + Sync + Send,
+    H: HashFunctionType + Sync + Send,
+    LF: LeafFact<S, H> + Sync + Send,
+{
+    fn is_leaf(&self) -> bool {
+        self.get_height_in_tree() == Height(0)
+    }
+    fn _leaf_hash(&self) -> Vec<u8>;
+    fn leaf_hash(&self) -> Option<Vec<u8>> {
+        match self.is_leaf() {
+            true => Some(self._leaf_hash()),
+            false => None,
+        }
+    }
+
+    /// Returns the height of the node in a tree.
+    fn get_height_in_tree(&self) -> Height;
+
+    fn create_leaf(hash_value: Vec<u8>) -> Self;
+
+    /// Returns the two BinaryFactTreeNode objects which are the roots of the subtrees of the
+    /// current BinaryFactTreeNode.
+    ///
+    /// If facts argument is not None, this dictionary is filled with facts read from the DB.
+    fn get_children<'a>(
+        &'a self,
+        ffc: &'a mut FactFetchingContext<S, H>,
+        facts: &'a mut Option<BinaryFactDict>,
+    ) -> impl std::future::Future<Output = Result<(Self, Self), TreeError>> + Send;
+
+    fn _get_leaves<'a>(
```

Comment:
is the `_foo()` convention here simply to keep things as close to the Python implementation as possible? Seems this is implied by not being `pub` in rust, but I also see value in keeping things identical for now

Reply:
Oh, this is a `trait`... :thinking:

Reply:
Yeah the problem I have is that it's the custom implementation of a method that is then called by a method called `get_children()` with default implementation that relies on `_get_children()`. It's the type of stuff I'd clean up later.